server/middleware/auth/authentication.go

```diff
@@ -0,0 +1,226 @@
+	return &AuthenticationMiddleware{
+		service: svc,
+		url:     make(map[string][]string),
+	}
+}
+
+func (amw *AuthenticationMiddleware) AddURL(path string, methods ...string) {
+	// Router is only used to convert path to regex.
+	pathRegex, err := mux.NewRouter().Path(path).Methods(http.MethodGet).GetPathRegexp()
+	if err != nil {
+		panic(err)
+	}
+	m, ok := amw.url[pathRegex]
+	if !ok {
+		amw.url[pathRegex] = methods
+	} else {
+		amw.url[pathRegex] = append(m, methods...)
+	}
+}
+
+// Middleware supports signature authentication and responds to the client by generating information necessary for authentication.
+func (amw *AuthenticationMiddleware) Middleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// if it is not an authentication URL, the authentication check is not performed.
+		if !amw.isAuthenticationURL(r) {
+			next.ServeHTTP(w, r)
+			return
+		}
+
+		sigAuthParts, err := ParseSignatureAuthorizationParts(r.Header.Get("Authorization"))
+		//signatureAuthentication, err := parseSignatureAuthentication(r)
+		if err != nil {
+			// result code to 400. Bad request
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		err = basicValidate(sigAuthParts)
+		if err != nil {
+			// result code to 400. Bad request
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		if sigAuthParts[Nonce] == "" {
+			err = amw.generateAuthenticationAndSetHeader(w, sigAuthParts)
+			if err != nil {
+				log.Error("failed to generate authentication", err)
+				http.Error(w, "failed to generate authentication", http.StatusInternalServerError)
+			}
+			http.Error(w, "", http.StatusUnauthorized)
+			return
+		}
+
+		cache := amw.service.Cache
+		auth := cache.Get(sigAuthParts[KeyId], sigAuthParts[Nonce])
+		if auth == nil {
+			// expired
+			err = amw.generateAuthenticationAndSetHeader(w, sigAuthParts)
+			if err != nil {
+				log.Error("failed to generate authentication", err)
+				http.Error(w, "failed to generate authentication", http.StatusInternalServerError)
+			}
+			http.Error(w, "", http.StatusUnauthorized)
+			return
+		}
+
+		nonce := sigAuthParts[Nonce]
+		signature, err := base64.StdEncoding.DecodeString(sigAuthParts[Signature])
+		if err != nil {
+			http.Error(w, "failed to decode signature", http.StatusBadRequest)
+			return
+		}
+
+		requesterAddress := sigAuthParts[KeyId]
+		pubKey, err := amw.service.PanaceaClient.GetPubKey(requesterAddress)
+		if err != nil {
+			log.Error(fmt.Sprintf("failed to get the account's public key account: %s", requesterAddress), err)
+			http.Error(w, "failed to get the account's public key", http.StatusInternalServerError)
+			return
+		}
+
+		if !pubKey.VerifySignature([]byte(nonce), signature) {
+			http.Error(w, "failed to verification signature", http.StatusBadRequest)
+			return
+		}
+
+		context.Set(r, types.RequesterAddressKey, requesterAddress)
```

Comment:
That's right. It can manipulate headers.
I think the header can also be received from the client, so it can be manipulated.
Also, since we are open source, the client can know which header to write.

Of course we could add logic to defend against this header.
I didn't want to touch the client's header 😀

Reply:
As you said, any malicious can manipulate the header, so I agree not to use requester header. Thanks!